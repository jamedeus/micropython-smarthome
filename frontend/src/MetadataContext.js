import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';
import { parse_dom_context } from 'util/django_util';

export const MetadataContext = createContext();

export const MetadataContextProvider = ({ children }) => {
    // Takes metadata entry, replaces "placeholder" with "" in config template
    // TODO probably not necessary to use placeholder in metadata anyway
    // Think only config_generator.py relies on them, can just check for "" instead
    const remove_placeholders = (template) => {
        for (let param in template.config_template) {
            if (template.config_template[param] === "placeholder") {
                template.config_template[param] = "";
            }
        }
        return template;
    };

    // Get device/sensor metadata object, contains rule types and config templates
    // Edit_config appends config template to state when device/sensor type selected
    // Rule prompts and limits used to render correct inputs, set slider ranges, etc
    const [metadata] = useState(() => {
        const metadata = parse_dom_context("instance_metadata");

        /* istanbul ignore else */
        if (metadata) {
            // Remove "placeholder" string from config templates in metadata object
            for (let device in metadata.devices) {
                metadata.devices[device] = remove_placeholders(metadata.devices[device]);
            }
            for (let sensor in metadata.sensors) {
                metadata.sensors[sensor] = remove_placeholders(metadata.sensors[sensor]);
            }
        }

        return metadata;
    });

    // Takes category ("device" or "sensor") and type, returns full metadata object
    // Creates deep copy so modifying object does not affect other components
    const get_instance_metadata = (category, type) => {
        return { ...metadata[`${category}s`][type],
            config_template: get_config_template(category, type)
        };
    };

    // Takes category ("device" or "sensor") and type, returns config template
    // Creates copy so modifying template does not affect other components
    const get_config_template = (category, type) => {
        if (category && type) {
            return { ...metadata[`${category}s`][type]['config_template'] };
        } else {
            return null;
        }
    };

    return (
        <MetadataContext.Provider value={{ metadata, get_instance_metadata }}>
            {children}
        </MetadataContext.Provider>
    );
};

MetadataContextProvider.propTypes = {
    children: PropTypes.node,
};
