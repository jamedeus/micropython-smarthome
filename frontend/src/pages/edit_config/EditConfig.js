import React, { useState, useContext } from 'react';
import { ConfigContext } from 'root/ConfigContext';
import Button from 'react-bootstrap/Button';
import { send_post_request, edit_existing, target_node_ip } from 'util/django_util';
import Page1 from './Page1';
import Page2 from './Page2';
import Page3 from './Page3';
import ApiTargetRuleModal from 'modals/ApiTargetRuleModal';
import { ErrorModalContext, ErrorModal } from 'modals/ErrorModal';
import UploadModal, { uploadConfigFile } from 'modals/UploadModal';

// Redirect back to overview page
function returnToOverview() {
    window.location.replace("/config_overview");
}

// Takes current config (state) object, compares with original from django
// context, returns True if changes have been made, false if no changes
function configModified(config) {
    // Get original config, compare with
    const original_config = JSON.parse(
        document.getElementById("config").textContent
    );
    return !areObjectsEqual(config, original_config);
}

// Takes 2 objects, recursively compares all subkeys, returns True if identical
// Used to detect unsaved changes, show warning before returning to overview
function areObjectsEqual(obj1, obj2) {
    // Direct comparison for sub-keys on recursive calls
    if (obj1 === obj2) {
        return true;
    }

    // Not equal if failed comparison above and either is not object
    if (typeof(obj1) !== 'object' || typeof(obj2) !== 'object') {
        return false;
    }

    const keys1 = Object.keys(obj1);
    const keys2 = Object.keys(obj2);

    // Not equal if different number of keys
    if (keys1.length !== keys2.length) {
        return false;
    }

    // Confirm contain same keys
    // Call recursively with values of each key in both objects
    for (let key of keys1) {
        if (!keys2.includes(key) || !areObjectsEqual(obj1[key], obj2[key])) {
            return false;
        }
    }

    return true;
}

// Takes config (state) object, returns true if any keys are empty strings
// Used to detect empty fields before changing page
function hasEmptyFields(config) {
    for (let key in config) {
        if (typeof(config[key]) === 'object') {
            if (hasEmptyFields(config[key])) {
                console.log(`Empty field in ${key}`);
                return true;
            }
        } else if ([null, '', undefined].includes(config[key])) {
            console.log(`Empty field: ${key}`);
            return true;
        }
    }
    return false;
}

// Takes config object, returns true if any keys are null or empty strings
// Used to detect unset schedule rule time before submitting
function hasEmptyKeys(config) {
    for (let key in config) {
        if ([null, '', undefined].includes(key)) {
            console.log('Config contains empty key');
            return true;
        } else if (typeof(config[key]) === 'object') {
            if (hasEmptyKeys(config[key])) {
                console.log('Config contains empty key');
                return true;
            }
        }
    }
}

const EditConfig = () => {
    // Set default page, get callback to change visible page
    const [page, setPage] = useState(1);

    // Get full config (state object)
    const { config, setHighlightInvalid } = useContext(ConfigContext);

    // Get state and callback for error modal
    const {
        errorModalContent,
        setErrorModalContent
    } = useContext(ErrorModalContext);

    function prevPage() {
        // Go back to overview if current page is page 1
        if (page === 1) {
            // Show unsaved changes warning if user modified any inputs
            if (configModified(config)) {
                setErrorModalContent({
                    ...errorModalContent,
                    ["visible"]: true,
                    ["title"]: "Warning",
                    ["error"]: "unsaved_changes",
                    ["handleConfirm"]: returnToOverview
                });
            // Go directly to overview if no unsaved changes
            } else {
                returnToOverview();
            }
        // Otherwise go to previous page
        } else {
            setPage(page - 1);
        }
    }

    function nextPage() {
        // Don't go to page2 if empty inputs exist on page1
        if (page === 1 && hasEmptyFields(config)) {
            setHighlightInvalid(true);
            return;
        }
        // Clear highlight, go to next page
        setHighlightInvalid(false);
        setPage(page + 1);
    }

    function friendlyNameToFilename(friendlyName) {
        return `${friendlyName.toLowerCase().replaceAll(' ', '-')}.json`;
    }

    // Post full config (state object) to backend when submit clicked
    async function submitButton() {
        console.log(config);

        // Don't submit if config has empty keys (schedule rule time not set)
        if (hasEmptyKeys(config)) {
            setHighlightInvalid(true);
            return;
        }

        // Overwrites if editing existing config, otherwise create config
        const response = await send_post_request(
            edit_existing ? "generate_config_file/True" : "generate_config_file",
            config
        );

        // If successfully created new config, redirect to overview
        if (!edit_existing && response.ok) {
            // Redirect back to overview where user can upload the new config
            returnToOverview();

        // If successfully edited existing config, re-upload to target node
        } else if (edit_existing && response.ok) {
            // Convert friendly name into config filename
            const target_filename = friendlyNameToFilename(config.metadata.id);
            // Show upload modal, upload, redirect to overview when complete
            uploadConfigFile(target_filename, target_node_ip, true, returnToOverview);

        // If config with same name already exists, show overwrite prompt
        } else if (!edit_existing && response.status == 409) {
            setErrorModalContent({
                ...errorModalContent,
                ["visible"]: true,
                ["title"]: "Duplicate Warning",
                ["error"]: "duplicate",
                ["body"]: config.metadata.id,
                ["handleConfirm"]: confirmOverwriteDuplicate
            });

        // If other error, display in alert
        } else {
            alert(await response.text());
            setHighlightInvalid(true);
        }
    }

    // Handler for error modal overwrite button when showing duplicate error
    async function confirmOverwriteDuplicate() {
        // Convert friendly name into config filename
        const target_filename = friendlyNameToFilename(config.metadata.id);
        // Close error modal, delete existing file, resubmit
        setErrorModalContent({ ...errorModalContent, ["visible"]: false });
        await send_post_request("delete_config", `${target_filename}.json`);
        await submitButton();
    }

    const PrevPageButton = () => {
        return (
            <Button variant="primary" className="mb-4" onClick={prevPage}>
                Back
            </Button>
        );
    };

    const SubmitButton = () => {
        return (
            <Button variant="primary" className="mb-4" onClick={submitButton}>
                Submit
            </Button>
        );
    };

    const NextPageButton = () => {
        return (
            <Button
                variant="primary"
                className="mb-4"
                onClick={nextPage}
                disabled={page === 3}
            >
                Next
            </Button>
        );
    };

    return (
        <>
            <div className="d-flex flex-column vh-100">
                <h1 className="text-center pt-3 pb-4">{document.title}</h1>
                {(() => {
                    switch(page) {
                        case 1:
                            return <Page1 />;
                        case 2:
                            return <Page2 />;
                        case 3:
                            return <Page3 />;
                    }
                })()}
                <div className="d-flex justify-content-between mx-3 mt-auto">
                    <PrevPageButton />
                    {page === 3 ? <SubmitButton /> : null}
                    <NextPageButton />
                </div>
            </div>
            <ApiTargetRuleModal />
            <UploadModal />
            <ErrorModal />
        </>
    );
};

export default EditConfig;
