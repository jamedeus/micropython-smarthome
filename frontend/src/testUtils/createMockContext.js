// Creates DOM elements containing JSON (mocks django json_script filter)
// Takes context id (arg passed to parse_dom_context) and json contents
const createMockContext = (id, json) => {
    const mockContext = document.createElement('div');
    mockContext.id = id;
    mockContext.textContent = JSON.stringify(json);
    document.body.appendChild(mockContext);
};

export default createMockContext;
