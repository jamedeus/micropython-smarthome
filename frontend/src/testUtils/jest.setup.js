import { render, within } from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import { act, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

beforeAll(() => {
    // Mock methods not implemented in jsdom
    window.HTMLElement.prototype.scrollIntoView = jest.fn();

    // Make available in all tests
    global.render = render;
    global.within = within;
    global.userEvent = userEvent;
    global.act = act;
    global.waitFor = waitFor;

    // Mock method called when window.location.href set
    Object.defineProperty(window, 'location', {
        value: {
            assign: jest.fn(),
        },
    });

    global.history.pushState = jest.fn();

    // Silence error (maybe from chart.js?)
    HTMLCanvasElement.prototype.getContext = jest.fn();
});

beforeEach(() => {
    // Reset number of calls for each mock to isolate test
    jest.clearAllMocks();
});
