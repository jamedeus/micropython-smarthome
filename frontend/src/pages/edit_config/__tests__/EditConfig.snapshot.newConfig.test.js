import React from 'react';
import EditConfig from '../EditConfig';
import { ConfigProvider } from 'root/ConfigContext';
import { MetadataContextProvider } from 'root/MetadataContext';
import createMockContext from 'src/testUtils/createMockContext';
import { newConfigContext, apiTargetOptionsContext } from './mockContext';
import { edit_config_metadata } from 'src/testUtils/mockMetadataContext';

describe('App', () => {
    it('matches snapshot when creating new config file', () => {
        // Create mock state objects
        createMockContext('config', newConfigContext.config);
        createMockContext('api_target_options', apiTargetOptionsContext);
        createMockContext('instance_metadata', edit_config_metadata);
        createMockContext('edit_existing', newConfigContext.edit_existing);

        // Render App, confirm matches snapshot
        const component = render(
            <MetadataContextProvider>
                <ConfigProvider>
                    <EditConfig />
                </ConfigProvider>
            </MetadataContextProvider>
        );
        expect(component).toMatchSnapshot();
    });
});
