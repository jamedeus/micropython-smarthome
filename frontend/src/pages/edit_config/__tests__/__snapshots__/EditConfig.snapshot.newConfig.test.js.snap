// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot when creating new config file 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="config"
    >
      {"metadata":{"id":"","floor":"","location":"","schedule_keywords":{"morning":"08:00","sleep":"23:00","sunrise":"06:00","sunset":"18:00","relax":"20:00"}}}
    </div>
    <div
      id="api_target_options"
    >
      {"addresses":{"self-target":"127.0.0.1","Bathroom":"192.168.1.100","Kitchen":"192.168.1.101","Living Room":"192.168.1.102","Bedroom":"192.168.1.103","Thermostat":"192.168.1.104"},"self-target":{},"Bathroom":{"sensor1":{"display":"Motion Sensor (pir)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"device1":{"display":"Bathroom LEDs (pwm)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"device2":{"display":"Bathroom Lights (tasmota-relay)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"ignore":{"display":"Ignore action"}},"Kitchen":{"sensor1":{"display":"Motion Sensor (pir)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"device1":{"display":"Cabinet Lights (pwm)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"device2":{"display":"Overhead Lights (tasmota-relay)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"ignore":{"display":"Ignore action"}},"Living Room":{"sensor1":{"display":"Motion Sensor (pir)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"sensor2":{"display":"Sunset Sensor (dummy)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"device1":{"display":"Overhead Lights (dimmer)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"device2":{"display":"Lamp (bulb)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"device3":{"display":"Porch Light (tasmota-relay)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"ignore":{"display":"Ignore action"}},"Thermostat":{"device1":{"display":"Air Conditioner (api-target)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"sensor1":{"display":"Thermostat (si7021)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule"]},"ir_key":{"display":"Ir Blaster","options":["tv","ac"],"keys":{"tv":["power","vol_up","vol_down","mute","up","down","left","right","enter","settings","exit","source"],"ac":["start","stop","off"]}},"ignore":{"display":"Ignore action"}},"Bedroom":{"sensor1":{"display":"Motion Sensor (pir)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"device1":{"display":"Lights (dimmer)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"device2":{"display":"Bias lights (wled)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","turn_on","turn_off"]},"sensor2":{"display":"Override (dummy)","options":["enable","disable","enable_in","disable_in","set_rule","reset_rule","trigger_sensor"]},"ignore":{"display":"Ignore action"}}}
    </div>
    <div
      id="instance_metadata"
    >
      {"devices":{"mosfet":{"config_name":"mosfet","class_name":"Mosfet","display_name":"Mosfet","dependencies":["devices/Mosfet.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"mosfet","nickname":"placeholder","default_rule":"placeholder","pin":"placeholder","schedule":{}},"rule_prompt":"standard"},"http-get":{"config_name":"http-get","class_name":"HttpGet","display_name":"HTTP Get Request","dependencies":["devices/HttpGet.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"http-get","nickname":"placeholder","default_rule":"placeholder","uri":"placeholder","on_path":"placeholder","off_path":"placeholder","schedule":{}},"rule_prompt":"standard"},"pwm":{"config_name":"pwm","class_name":"LedStrip","display_name":"LED Strip","dependencies":["devices/LedStrip.py","devices/DimmableLight.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"pwm","nickname":"placeholder","min_rule":"placeholder","max_rule":"placeholder","default_rule":"placeholder","pin":"placeholder","schedule":{}},"rule_prompt":"int_or_fade","rule_limits":[0,1023]},"desktop":{"config_name":"desktop","class_name":"DesktopTarget","display_name":"Computer Screen","dependencies":["devices/DesktopTarget.py","devices/HttpGet.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"desktop","nickname":"placeholder","ip":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"standard"},"tasmota-relay":{"config_name":"tasmota-relay","class_name":"TasmotaRelay","display_name":"Tasmota Smart Relay","dependencies":["devices/TasmotaRelay.py","devices/HttpGet.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"tasmota-relay","nickname":"placeholder","ip":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"standard"},"dimmer":{"config_name":"dimmer","class_name":"Tplink","display_name":"TP Link Smart Dimmer","dependencies":["devices/Tplink.py","devices/DimmableLight.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"dimmer","nickname":"placeholder","ip":"placeholder","min_rule":"placeholder","max_rule":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"int_or_fade","rule_limits":[1,100]},"api-target":{"config_name":"api-target","class_name":"ApiTarget","display_name":"API command","dependencies":["devices/ApiTarget.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"api-target","nickname":"placeholder","ip":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"api_target"},"dumb-relay":{"config_name":"dumb-relay","class_name":"DumbRelay","display_name":"Relay","dependencies":["devices/DumbRelay.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"dumb-relay","nickname":"placeholder","default_rule":"placeholder","pin":"placeholder","schedule":{}},"rule_prompt":"standard"},"bulb":{"config_name":"bulb","class_name":"Tplink","display_name":"TP Link Smart Bulb","dependencies":["devices/Tplink.py","devices/DimmableLight.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"bulb","nickname":"placeholder","ip":"placeholder","min_rule":"placeholder","max_rule":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"int_or_fade","rule_limits":[1,100]},"wled":{"config_name":"wled","class_name":"Wled","display_name":"WLED","dependencies":["devices/Wled.py","devices/DimmableLight.py","devices/Device.py","core/Instance.py"],"config_template":{"_type":"wled","nickname":"placeholder","ip":"placeholder","min_rule":"placeholder","max_rule":"placeholder","default_rule":"placeholder","schedule":{}},"rule_prompt":"int_or_fade","rule_limits":[1,255]}},"sensors":{"switch":{"config_name":"switch","class_name":"Switch","display_name":"Switch","dependencies":["sensors/Switch.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"switch","nickname":"placeholder","default_rule":"placeholder","pin":"placeholder","schedule":{},"targets":[]},"rule_prompt":"standard","triggerable":false},"dht22":{"config_name":"dht22","class_name":"Dht22","display_name":"DHT22 Temperature Sensor","dependencies":["sensors/Dht22.py","sensors/Thermostat.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"dht22","nickname":"placeholder","units":"placeholder","default_rule":"placeholder","mode":"placeholder","tolerance":"placeholder","pin":"placeholder","schedule":{},"targets":[]},"rule_prompt":"float_range","rule_limits":[18,27],"triggerable":false},"si7021":{"config_name":"si7021","class_name":"Si7021","display_name":"SI7021 Temperature Sensor","dependencies":["sensors/Si7021.py","sensors/Thermostat.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"si7021","nickname":"placeholder","units":"placeholder","default_rule":"placeholder","mode":"placeholder","tolerance":"placeholder","schedule":{},"targets":[]},"rule_prompt":"float_range","rule_limits":[18,27],"triggerable":false},"desktop":{"config_name":"desktop","class_name":"DesktopTrigger","display_name":"Computer Activity","dependencies":["sensors/DesktopTrigger.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"desktop","nickname":"placeholder","default_rule":"placeholder","ip":"placeholder","schedule":{},"targets":[]},"rule_prompt":"standard","triggerable":true},"pir":{"config_name":"pir","class_name":"MotionSensor","display_name":"PIR Motion Sensor","dependencies":["sensors/MotionSensor.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"pir","nickname":"placeholder","default_rule":"placeholder","pin":"placeholder","schedule":{},"targets":[]},"rule_prompt":"float_range","rule_limits":[0,60],"triggerable":true},"load-cell":{"config_name":"load-cell","class_name":"LoadCell","display_name":"Load Cell Pressure Sensor","dependencies":["sensors/LoadCell.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"load-cell","nickname":"placeholder","default_rule":"placeholder","pin_data":"placeholder","pin_clock":"placeholder","schedule":{},"targets":[]},"rule_prompt":"float_range","rule_limits":[0,10000000],"triggerable":false},"dummy":{"config_name":"dummy","class_name":"Dummy","display_name":"Dummy Sensor","dependencies":["sensors/Dummy.py","sensors/Sensor.py","core/Instance.py"],"config_template":{"_type":"dummy","nickname":"placeholder","default_rule":"placeholder","schedule":{},"targets":[]},"rule_prompt":"on_off","triggerable":true}}}
    </div>
    <div
      id="edit_existing"
    >
      false
    </div>
    <div
      id="ir_blaster_targets"
    >
      ["samsung","whynter"]
    </div>
    <div>
      <div
        class="d-flex flex-column vh-100"
      >
        <h1
          class="text-center pt-3 pb-4"
        />
        <div
          id="metadata"
        >
          <div
            class="mb-4 mx-3"
          >
            <div
              class="mb-2"
            >
              <label
                class="w-100 fw-bold"
              >
                Name
                :
                <input
                  class="form-control"
                  type="text"
                  value=""
                />
                <div
                  class="invalid-feedback"
                >
                  Name must be unique
                </div>
              </label>
            </div>
            <div
              class="mb-2"
            >
              <label
                class="w-100 fw-bold"
              >
                Location
                :
                <input
                  class="form-control"
                  type="text"
                  value=""
                />
              </label>
            </div>
            <div
              class="mb-2"
            >
              <label
                class="w-100 fw-bold"
              >
                Floor
                :
                <input
                  class="form-control"
                  type="text"
                  value=""
                />
              </label>
            </div>
          </div>
        </div>
        <div
          class="max-width-md-50 w-100 mx-auto text-center"
        >
          <p
            class="text-center mt-3"
          >
            <button
              class="btn btn-secondary"
              type="button"
            >
              Add IR Blaster
            </button>
          </p>
          <div
            class="collapse"
          >
            <div
              class="card"
            >
              <div
                class="mx-auto card-body"
              >
                <h2>
                  IR Blaster
                </h2>
                <div
                  class="mb-2"
                >
                  <label
                    class="w-100 fw-bold"
                  >
                    Pin
                    :
                    <select
                      class="form-select"
                    >
                      <option
                        value=""
                      >
                        Select pin
                      </option>
                      <option
                        value="4"
                      >
                        4
                      </option>
                      <option
                        value="13"
                      >
                        13
                      </option>
                      <option
                        value="16"
                      >
                        16
                      </option>
                      <option
                        value="17"
                      >
                        17
                      </option>
                      <option
                        value="18"
                      >
                        18
                      </option>
                      <option
                        value="19"
                      >
                        19
                      </option>
                      <option
                        value="21"
                      >
                        21
                      </option>
                      <option
                        value="22"
                      >
                        22
                      </option>
                      <option
                        value="23"
                      >
                        23
                      </option>
                      <option
                        value="25"
                      >
                        25
                      </option>
                      <option
                        value="26"
                      >
                        26
                      </option>
                      <option
                        value="27"
                      >
                        27
                      </option>
                      <option
                        value="32"
                      >
                        32
                      </option>
                      <option
                        value="33"
                      >
                        33
                      </option>
                    </select>
                  </label>
                </div>
                <div
                  class="mb-2"
                >
                  <label
                    class="fw-bold"
                    for="ir-remotes"
                  >
                    Virtual remotes:
                  </label>
                  <div
                    id="ir-remotes"
                  >
                    <div
                      class="form-check"
                    >
                      <input
                        class="form-check-input"
                        id="samsung-codes"
                        type="checkbox"
                      />
                      <label
                        class="form-check-label"
                        for="samsung-codes"
                        title=""
                      >
                        Samsung
                      </label>
                    </div>
                    <div
                      class="form-check"
                    >
                      <input
                        class="form-check-input"
                        id="whynter-codes"
                        type="checkbox"
                      />
                      <label
                        class="form-check-label"
                        for="whynter-codes"
                        title=""
                      >
                        Whynter
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-flex flex-column"
        >
          <div
            class="mt-3 row"
          >
            <div
              class="col-sm"
            >
              <h2
                class="text-center"
              >
                Add 
                Sensor
                s
              </h2>
              <div
                class="text-center position-relative mb-3 "
              >
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Add 
                  Sensor
                </button>
              </div>
            </div>
            <div
              class="col-sm"
            >
              <h2
                class="text-center"
              >
                Add 
                Device
                s
              </h2>
              <div
                class="text-center position-relative mb-3 "
              >
                <button
                  class="btn btn-secondary"
                  type="button"
                >
                  Add 
                  Device
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-flex justify-content-between mx-3 mt-auto"
        >
          <button
            class="mb-4 btn btn-primary"
            type="button"
          >
            Back
          </button>
          <button
            class="mb-4 btn btn-primary"
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column vh-100"
    >
      <h1
        class="text-center pt-3 pb-4"
      />
      <div
        id="metadata"
      >
        <div
          class="mb-4 mx-3"
        >
          <div
            class="mb-2"
          >
            <label
              class="w-100 fw-bold"
            >
              Name
              :
              <input
                class="form-control"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              >
                Name must be unique
              </div>
            </label>
          </div>
          <div
            class="mb-2"
          >
            <label
              class="w-100 fw-bold"
            >
              Location
              :
              <input
                class="form-control"
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            class="mb-2"
          >
            <label
              class="w-100 fw-bold"
            >
              Floor
              :
              <input
                class="form-control"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
      </div>
      <div
        class="max-width-md-50 w-100 mx-auto text-center"
      >
        <p
          class="text-center mt-3"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            Add IR Blaster
          </button>
        </p>
        <div
          class="collapse"
        >
          <div
            class="card"
          >
            <div
              class="mx-auto card-body"
            >
              <h2>
                IR Blaster
              </h2>
              <div
                class="mb-2"
              >
                <label
                  class="w-100 fw-bold"
                >
                  Pin
                  :
                  <select
                    class="form-select"
                  >
                    <option
                      value=""
                    >
                      Select pin
                    </option>
                    <option
                      value="4"
                    >
                      4
                    </option>
                    <option
                      value="13"
                    >
                      13
                    </option>
                    <option
                      value="16"
                    >
                      16
                    </option>
                    <option
                      value="17"
                    >
                      17
                    </option>
                    <option
                      value="18"
                    >
                      18
                    </option>
                    <option
                      value="19"
                    >
                      19
                    </option>
                    <option
                      value="21"
                    >
                      21
                    </option>
                    <option
                      value="22"
                    >
                      22
                    </option>
                    <option
                      value="23"
                    >
                      23
                    </option>
                    <option
                      value="25"
                    >
                      25
                    </option>
                    <option
                      value="26"
                    >
                      26
                    </option>
                    <option
                      value="27"
                    >
                      27
                    </option>
                    <option
                      value="32"
                    >
                      32
                    </option>
                    <option
                      value="33"
                    >
                      33
                    </option>
                  </select>
                </label>
              </div>
              <div
                class="mb-2"
              >
                <label
                  class="fw-bold"
                  for="ir-remotes"
                >
                  Virtual remotes:
                </label>
                <div
                  id="ir-remotes"
                >
                  <div
                    class="form-check"
                  >
                    <input
                      class="form-check-input"
                      id="samsung-codes"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="samsung-codes"
                      title=""
                    >
                      Samsung
                    </label>
                  </div>
                  <div
                    class="form-check"
                  >
                    <input
                      class="form-check-input"
                      id="whynter-codes"
                      type="checkbox"
                    />
                    <label
                      class="form-check-label"
                      for="whynter-codes"
                      title=""
                    >
                      Whynter
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex flex-column"
      >
        <div
          class="mt-3 row"
        >
          <div
            class="col-sm"
          >
            <h2
              class="text-center"
            >
              Add 
              Sensor
              s
            </h2>
            <div
              class="text-center position-relative mb-3 "
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Add 
                Sensor
              </button>
            </div>
          </div>
          <div
            class="col-sm"
          >
            <h2
              class="text-center"
            >
              Add 
              Device
              s
            </h2>
            <div
              class="text-center position-relative mb-3 "
            >
              <button
                class="btn btn-secondary"
                type="button"
              >
                Add 
                Device
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex justify-content-between mx-3 mt-auto"
      >
        <button
          class="mb-4 btn btn-primary"
          type="button"
        >
          Back
        </button>
        <button
          class="mb-4 btn btn-primary"
          type="button"
        >
          Next
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
